src/arm7tdmi.rs

```diff
@@ -29,40 +35,90 @@ impl Cpu for Arm7tdmi {
         op_code
     }
 
-    fn decode(&self, op_code: Self::OpCodeType) -> (Condition, InstructionKind) {
+    fn decode(&self, op_code: Self::OpCodeType) -> (Condition, Self::InstructionType) {
+
         let condition = (op_code >> 28) as u8; // bit 31..=28
-        let instruction = if (op_code & 0x0A_00_00_00) != 0 {
-            if (op_code & 0x01_00_00_00) != 0 {
-                InstructionKind::BranchLink
-            } else {
-                InstructionKind::Branch
-            }
-        } else {
-            todo!()
-        };
         println!("condition -> {:x}", condition);
+
+        let res_decode = ArmModeInstruction::get_instruction(op_code);
+        if res_decode.is_err() {
+            todo!("ArmModeInstruction")
+        }
+        let instruction = res_decode.expect("ArmMode");
         println!("instruction -> {:?}", instruction);
+
         (condition.into(), instruction)
     }
 
-    fn execute(&self) {}
+    fn execute(&mut self, op_code: u32, instruction_type: ArmModeInstruction) {
+        use ArmModeInstruction::*;
+        match instruction_type {
+            Branch => {
+                self.branch(op_code);
+            }
+            BranchLink => {
+                self.branch_link(op_code);
+            }
+            _ => todo!("Instruction not implemented yet.")
+        }
+    }
 
-    fn step(&self) {
+    fn step(&mut self) {
         let op_code = self.fetch();
 
-        let condition = self.decode(op_code);
-        if self.cpsr.can_execute(condition.0) {
-            todo!("we can execute now")
+        let (condition,instruction) = self.decode(op_code);
+        if self.cpsr.can_execute(condition) {
+            self.execute(op_code, instruction)
         }
     }
 }
 
 impl Arm7tdmi {
-    pub(crate) fn new(data: Vec<u8>) -> Self {
+    pub(crate) fn new(rom: Vec<u8>) -> Self {
         Self {
-            data,
+            rom,
             program_counter: 0,
             cpsr: Cpsr::default(),
         }
     }
+
+    fn branch(&mut self, op_code: u32) {
+        let offset = op_code & 0x00_FF_FF_FF;
+        println!("offset: {:?}", offset);
+
+        self.program_counter += 8 + offset * 4;
+        println!("PC: {:?}", self.program_counter);
+    }
+
+    fn branch_link(&mut self, op_code: u32) {
+        todo!("Branch Link")
+    }
 }
+
+#[repr(u32)]
+#[derive(Debug)]
+pub(crate) enum ArmModeInstruction {
+    Branch = 0x0A_00_00_00,
+    BranchLink =  0x0B_00_00_00,
+}
+
+impl ArmModeInstruction {
+
+    fn get_instruction(op_code: u32) -> Result<ArmModeInstruction, Error> {
+        use ArmModeInstruction::*;
+        
+        if Self::check(Branch as u32, op_code) {
```

Comment:
We can create a function for `u32` like `op_code.equals(Branch)` but we can improve this in separate PR

Reply:
Yes, we can do it.